File: 2/2.go
package main

import (
	"github.com/lillrurre/adventofcode-2023/util"
	"regexp"
	"strings"
)

type game struct {
	red, green, blue int
}

func main() {
	s := util.FileAsScanner(2)
	re := regexp.MustCompile(`(\d+) (\w+)`)

	util.RunBoth(func() (int, int) {
		gameNum, part1, part2 := 1, 0, 0
		for s.Scan() {
			g := new(game)
			for _, match := range re.FindAllStringSubmatch(strings.TrimSpace(s.Text()), -1) {
				n := util.Atoi(match[1])
				color := match[2]
				switch color {
				case "red":
					g.red = max(g.red, n)
				case "green":
					g.green = max(g.green, n)
				case "blue":
					g.blue = max(g.blue, n)
				}
			}
			if g.red <= 12 && g.green <= 13 && g.blue <= 14 {
				part1 += gameNum
			}
			gameNum++

			part2 += g.red * g.green * g.blue
		}
		return part1, part2
	})
}

File: 2/2.rb
class Game
  attr_accessor :red, :green, :blue

  def initialize
    @red = 0
    @green = 0
    @blue = 0
  end
end

input = File.read("input/2")
re = /(\d+) (\w+)/

part1 = 0
part2 = 0

input.strip.split("\n").each_with_index do |s, i|
  g = Game.new
    s.scan(re).each do |match|
    n = match[0].to_i
    color = match[1]
    case
    when color == "red" && g.red < n
      g.red = n
    when color == "green" && g.green < n
      g.green = n
    when color == "blue" && g.blue < n
      g.blue = n
    end
  end

  if g.red <= 12 && g.green <= 13 && g.blue <= 14
    part1 += i + 1
  end
  part2 += g.red * g.green * g.blue
end

puts "[1] Result: #{part1}"
puts "[2] Result: #{part2}"